src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Helpers/RemoveUnnecessaryImports/AbstractUnnecessaryImportsProvider.cs

```diff
@@ -2,36 +2,29 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 
 namespace Microsoft.CodeAnalysis.RemoveUnnecessaryImports
 {
-    internal abstract class AbstractUnnecessaryImportsProvider<T>
-        : IUnnecessaryImportsProvider, IEqualityComparer<T> where T : SyntaxNode
+    internal abstract class AbstractUnnecessaryImportsProvider<TSyntaxNode> :
+        IUnnecessaryImportsProvider<TSyntaxNode>,
+        IEqualityComparer<TSyntaxNode>
+        where TSyntaxNode : SyntaxNode
     {
-        public ImmutableArray<SyntaxNode> GetUnnecessaryImports(
-            SemanticModel model, CancellationToken cancellationToken)
-        {
-            var root = model.SyntaxTree.GetRoot(cancellationToken);
-            return GetUnnecessaryImports(model, root, predicate: null, cancellationToken: cancellationToken);
-        }
-
-        protected abstract ImmutableArray<SyntaxNode> GetUnnecessaryImports(
-            SemanticModel model, SyntaxNode root,
-            Func<SyntaxNode, bool> predicate, CancellationToken cancellationToken);
+        public abstract ImmutableArray<TSyntaxNode> GetUnnecessaryImports(
+            SemanticModel model, Func<SyntaxNode, bool>? predicate, CancellationToken cancellationToken);
```

Comment:
simpler api shape.  no need to pass the root, it's implied by the SemanticModel.